plugins/spidmx/SpiDmxThread.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * SpiDmxThread.cpp
+ * This thread runs while one or more ports are registered. It simultaneously
+ * reads / writes SPI data and then calls the parser. This is repeated forever.
+ * Copyright (C) 2017 Florian Edelmann
+ */
+
+#include <string>
+#include "ola/Logging.h"
+#include "plugins/spidmx/SpiDmxWidget.h"
+#include "plugins/spidmx/SpiDmxThread.h"
+#include "plugins/spidmx/SpiDmxParser.h"
+
+namespace ola {
+namespace plugin {
+namespace spidmx {
+
+SpiDmxThread::SpiDmxThread(SpiDmxWidget *widget, unsigned int blocklength)
+  : m_widget(widget),
+    m_blocklength(blocklength),
+    m_term(false),
+    m_registered_ports(0),
+    m_spi_rx_buffer(blocklength),
+    m_spi_tx_buffer(blocklength) {
+}
+
+SpiDmxThread::~SpiDmxThread() {
+  Stop();
+}
+
+/**
+ * This thread does only have to run if ports using it are patched to a
+ * universe. Thus, they must register and unregister to notify this thread.
+ */
+void SpiDmxThread::RegisterPort() {
+  m_registered_ports++;
+
+  if (m_registered_ports == 1) {
```

Comment:
>= 1 perhaps, although it shouldn't happen.

Reply:
Well done for understanding what I meant, even when the GitHub markdown confused it.